LibVLCSharp/Shared/LibVLCCallbacks.cs

```diff
@@ -0,0 +1,491 @@
+﻿using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using LibVLCSharp.Shared.Helpers;
+
+namespace LibVLCSharp.Shared
+{
+    /// <summary>
+    /// A static class that contains callbacks that libvlc can call.
+    ///
+    /// In many platforms, we can use any method or lambda as a native callback,
+    /// but not on Apple platforms, which requires static methods, decorated with [MonoPInvokeCallback]
+    ///
+    /// This class declares all those callbacks, and each callback routes the event to the appropriate instance
+    /// thanks to the user data, which should contain the GCHandle of the instance to call.
+    /// The callback then casts the instance to the appropriate type, and routes the call to the appropriate
+    /// internal method.
+    ///
+    /// To avoid the garbage collection of the callbacks, a reference is held on each method.
+    /// </summary>
+    internal static class LibVLCCallbacks
+    {
+        #region Exit
+        internal static readonly InternalExitCallback ExitCallbackHandle = ExitCallback;
+
+        [MonoPInvokeCallback(typeof(InternalExitCallback))]
+        private static void ExitCallback(IntPtr libVLCHandle)
+        {
+            var libVLC = GetInstance<LibVLC>(libVLCHandle);
+            libVLC?._exitCallback?.Invoke();
+        }
+        #endregion
+
+        #region Log
+        internal static readonly InternalLogCallback LogCallbackHandle = LogCallback;
+
+        [MonoPInvokeCallback(typeof(InternalLogCallback))]
+        private static void LogCallback(IntPtr libVLCHandle, LogLevel logLevel, IntPtr logContext, IntPtr format, IntPtr args)
+        {
+            var libVLC = GetInstance<LibVLC>(libVLCHandle);
+
+            try
+            {
+                var message = MarshalUtils.GetLogMessage(format, args);
+
+                LibVLC.GetLogContext(logContext, out var module, out var file, out var line);
+
+                Action logAction = () => libVLC._log?.Invoke(null, new LogEventArgs(logLevel, message, module, file, line));
```

Comment:
You can use a local function here

` void logAction() => libVLC._log?.Invoke(null, new LogEventArgs(logLevel, message, module, file, line));`